File: lt/src/q0143ReorderList.ts
import { ListNode } from "./util/ListNode";

export function reorderList(head: ListNode | null): void {
    if (!head) return;

    // Find the middle of the list.
    let slow = head,
        fast = head;
    while (fast.next && fast.next.next) {
        slow = slow.next!;
        fast = fast.next.next;
    }

    // Split the list into two halves.
    let secondHalf = slow.next;
    slow.next = null;

    // Reverse the second half.
    let prev = null,
        current = secondHalf;
    while (current) {
        const next = current.next;
        current.next = prev;
        prev = current;
        current = next;
    }
    secondHalf = prev;

    // Merge the two halves.
    let firstHalf: ListNode | null = head;
    while (firstHalf && secondHalf) {
        const nextFirstHalf: ListNode | null = firstHalf.next;
        const nextSecondHalf = secondHalf.next;

        firstHalf.next = secondHalf;
        secondHalf.next = nextFirstHalf;

        firstHalf = nextFirstHalf;
        secondHalf = nextSecondHalf;
    }
}

File: lg/src/questions/q0143_reorder_list.go
package questions

import (
	"github.com/hi-rustin/lg/src/utils"
)

func reorderList(head *utils.ListNode) {
	n := 0

	// Get the total length.
	current := head
	for current != nil {
		current = current.Next
		n += 1
	}

	if n == 1 {
		return // No need to reorder.
	}

	// FInd the middle position.
	middle := n / 2
	current = head
	for i := 0; i < middle; i++ {
		current = current.Next
	}

	// Revert the second half.
	var prev *utils.ListNode
	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}

	// Merge the two halves
	first, second := head, prev
	for second.Next != nil {
		tmp := first.Next
		first.Next = second
		first = tmp

		tmp = second.Next
		second.Next = first
		second = tmp
	}
}
